src/internal/m365/collection/drive/collections_tree.go

```diff
@@ -277,65 +277,94 @@ func (c *Collections) populateTree(
+		var (
+			pageCount     int
+			pageItemCount int
+			hadReset      bool
+			err           error
+		)
 
-	for page, reset, done := pager.NextPage(); !done; page, reset, done = pager.NextPage() {
-		if el.Failure() != nil {
-			break
-		}
+		// TODO(keepers): to end in a correct state, we'll eventually need to run this
+		// query multiple times over, until it ends in an empty change set.
+		pager := c.handler.EnumerateDriveItemsDelta(
+			ctx,
+			driveID,
+			currDeltaLink,
+			api.CallConfig{
+				Select: api.DefaultDriveItemProps(),
+			})
+
+		for page, reset, done := pager.NextPage(); !done; page, reset, done = pager.NextPage() {
+			if el.Failure() != nil {
+				return du, el.Failure()
+			}
 
-		if reset {
-			counter.Inc(count.PagerResets)
-			tree.reset()
-			c.resetStats()
-		}
+			if reset {
+				counter.Inc(count.PagerResets)
+				tree.reset()
+				c.resetStats()
 
-		err := c.enumeratePageOfItems(
-			ctx,
-			tree,
-			drv,
-			page,
-			limiter,
-			counter,
-			errs)
-		if err != nil {
-			if errors.Is(err, errHitLimit) {
+				pageCount = 0
+				pageItemCount = 0
+				hadReset = true
+			}
+
+			err = c.enumeratePageOfItems(
+				ctx,
+				tree,
+				drv,
+				page,
+				limiter,
+				counter,
+				errs)
+			if err != nil {
+				if errors.Is(err, errHitLimit) {
+					hitLimit = true
+					break
+				}
+
+				el.AddRecoverable(ctx, clues.Stack(err))
+			}
+
+			counter.Inc(count.PagesEnumerated)
+
+			// Stop enumeration early if we've reached the page limit. Keep this
+			// at the end of the loop so we don't request another page (pager.NextPage)
+			// before seeing we've passed the limit.
+			if limiter.hitPageLimit(int(counter.Get(count.PagesEnumerated))) {
+				hitLimit = true
 				break
 			}
 
-			el.AddRecoverable(ctx, clues.Stack(err))
+			pageCount++
+
+			pageItemCount += len(page)
 		}
 
-		counter.Inc(count.PagesEnumerated)
+		// Always cancel the pager so that even if we exit early from the loop above
+		// we don't deadlock. Cancelling a pager that's already completed is
+		// essentially a noop.
+		pager.Cancel()
 
-		// Stop enumeration early if we've reached the page limit. Keep this
-		// at the end of the loop so we don't request another page (pager.NextPage)
-		// before seeing we've passed the limit.
-		if limiter.hitPageLimit(int(counter.Get(count.PagesEnumerated))) {
-			break
+		du, err = pager.Results()
+		if err != nil {
+			return du, clues.Stack(err)
 		}
-	}
 
-	// Always cancel the pager so that even if we exit early from the loop above
-	// we don't deadlock. Cancelling a pager that's already completed is
-	// essentially a noop.
-	pager.Cancel()
+		currDeltaLink = du.URL
 
-	du, err := pager.Results()
-	if err != nil {
-		return du, clues.Stack(err)
+		// 0 pages is never expected.  We should at least have one (empty) page to
+		// consume.  But checking pageCount == 1 is brittle in a non-helpful way.
+		finished = pageCount < 2 && pageItemCount == 0 && !hadReset
```

Comment:
I'm unsure if we actually need to check if there was a reset or not. Perhaps we should go clarify the interface provided by the pager at some point to make this more clear

From my current understanding, when a reset is seen in the pager, it will send the reset status as a result (possibly with other items from after the reset), reset the pager, and then continue enumerating. Assuming that's the behavior we want to keep in the interface going forward, it seems like there's a few outcomes we could get here based on whether the pager can return additional items in the same result as the reset

**items in same result, no items returned**
Here we'll get `hadReset = true`, `pageCount = 1`, `pageItemCount = 0`. In theory, running another delta should return one empty page but without the reset

**items in same result, some items**
Here we'll get `hadReset = true`, `pageCount = 1`, `pageItemCount > 0`. We'll set `finished = false` regardless of if we check `hadReset` or not because of `pageItemCount`

**items in next result**
I'm consolidating not getting items and getting items for this because in both cases we'll get `hadReset = true` and `pageCount >= 2` (one result for reset, one for the next query). In both those cases `finished = false` because of `pageCount` so we'll do another query anyway

Reply:
Unless I coded it wrong, the `hadReset` check only applies to that page, not to the full enumeration.   Checking the flag is to prevent `reset, no items -> items in next result` from escaping early, on the reset.  After that (unless another reset happens), `hadReset` should revert to false.

For all identifications of "any reset ever occurred", the tree tracks that as part of calling `tree.reset()`, and callers can compare `tree.hadReset` for the flag.

Reply:
D'oh, I grok now.  You're right, it's not a required check.